modules/redcap/php/configurations/redcapconfiguration.class.inc

```diff
@@ -0,0 +1,239 @@
+     * @param string $url  the instance url
+     *
+     * @return void
+     */
+    public function addInstance(string $name, string $url): void
+    {
+        if ($this->hasInstanceName($name)) {
+            throw new \LorisException(
+                "[redcap][config] instance '$name' already exists."
+            );
+        }
+        $this->_instances[$name] = new RedcapConfigurationInstance($name, $url);
+    }
+
+    /**
+     * Get a REDCap instance by name.
+     *
+     * @param string $instanceName the instance name
+     *
+     * @return ?RedcapConfigurationInstance a REDCap instance or null.
+     */
+    public function getInstanceByName(
+        string $instanceName
+    ): ?RedcapConfigurationInstance {
+        return $this->_instances[$instanceName] ?? null;
+    }
+
+    /**
+     * Get a REDCap instance by URL.
+     *
+     * @param string $instanceURL the instance URL
+     *
+     * @return ?RedcapConfigurationInstance a REDCap instance or null.
+     */
+    public function getInstanceByURL(
+        string $instanceURL
+    ): ?RedcapConfigurationInstance {
+        // normalize the given URL
+        $searchedURL = self::_normalizeURL($instanceURL);
+        $i           = null;
+        foreach ($this->_instances as $instance) {
+            // normalize instance URL
+            $iURL = self::_normalizeURL($instance->getURL());
+            // compare
+            if ($iURL === $searchedURL) {
+                $i = $instance;
+                break;
+            }
+        }
+        return $i;
+    }
+
+    /**
+     * Normalize URL by removing the trailing slash.
+     *
+     * @param string $url URL to normalize
+     *
+     * @return string normalized URL
+     */
+    private static function _normalizeURL(string $url): string
+    {
+        return str_ends_with($url, '/') ? substr($url, 0, strlen($url)-1) : $url;
+    }
+
+    /**
+     * Tells if a REDCap instance name exists.
+     *
+     * @param string $instanceName the instance name
+     *
+     * @return bool true if the instance exists, else false.
+     */
+    public function hasInstanceName(string $instanceName): bool
+    {
+        return !is_null($this->getInstanceByName($instanceName));
+    }
+
+    /**
+     * Add a REDCap proejct to a REDCap instance.
+     *
+     * @param string $instanceName a REDCap instance name
+     * @param string $projectID    a REDCap project ID
+     * @param string $projectToken a REDCap project token
+     *
+     * @throws \LorisException
+     *
+     * @return void
+     */
+    public function addProjectToInstance(
+        string $instanceName,
+        string $projectID,
+        string $projectToken
+    ): void {
+        if (!$this->hasInstanceName($instanceName)) {
+            throw new \LorisException(
+                "[redcap][config] instance '$instanceName' does not exist."
+            );
+        }
+        $this->_instances[$instanceName]->addProject($projectID, $projectToken);
```

Comment:
More idiomatic, this has one less look-up and plays better with nullability type checking:

```suggestion
    $instance = $this->getInstanceByName($instanceName);
    if ($instance === null) {
        throw new \LorisException(
            "[redcap][config] instance '$instanceName' does not exist."
        );
    }
    
    $instance->addProject($projectID, $projectToken);
```